### DISCUSSION TITLE
How to deal with row values

### DISCUSSION THREAD

**Discussion by prog110** [querydsl/querydsl#2939] [Q&A] [answered]
Is there a way to write queries that deal with row values like `WHERE (sale_date, sale_id) < (?, ?)` using querydsl?

```sql
SELECT *
  FROM sales
 WHERE (sale_date, sale_id) < (?, ?)
 ORDER BY sale_date DESC, sale_id DESC
 FETCH FIRST 10 ROWS ONLY
```

More info
https://use-the-index-luke.com/sql/partial-results/fetch-next-page

Earlier conversation from google group
https://groups.google.com/d/msg/querydsl/bG_2MssrcTo/MgcUpJ6UvaQJ

**Comment by prog110** [comment]
@timowest Can you share your views please

**Comment by Shredder121** [comment]
Not sure if that is possible, since `tuple()` then doesn't know about the target type (`Comparable`, etc.).
Maybe you'd be able to pull something off using templates, but your mileage may vary.

**Comment by andygoossens** [comment, marked as answer]
@thekalinga: You can use `Expressions.list(Expression<?>... exprs)` for this.

Example:
```
Expression<Tuple> leftTuple = Expressions.list(...);
Expression<Tuple> rightTuple = Expressions.list(...);

// then you need to compare the two...
Expressions.booleanOperation(Ops.LT, leftTuple, rightTuple);
```

But there are several issues with it:
* #2344: QueryDSL breaks when you pass just 1 parameter to the list method. But there is little point in creating a list with just one element anyway.
* It only worked for me for basic types like String, Boolean, Long, etc... e.g. As soon as I passed a Joda-Time's DateTime object the parameter binding went wrong. It was considering that parameter to be "VARBINARY" when it should have been "TIMESTAMP". But it seems you can work around this issue when you pass the same parameter value elsewhere in your query, outside the list. But then you probably want that part of the WHERE clause to be a no-op. It will then look something like this:
```
WHERE (sale_date = :sale_date OR 1 = 1) AND (sale_date, sale_id) < (:sale_date, :sale_id)
```

So it "sort of" works, but you need to fight a bit with QueryDSL to make it happen.
